README.md
rackspace_support Cookbook
===========================
Enables a node to be supported by Rackspace.  Installs common
tools Rackspace support finds useful when debugging system
problems.

Note: This cookbook uses the firewall cookbook and will attempt
to manage your iptables unless the node is rackconnected.

Requirements
------------
#### cookbooks
- `firewall`
- `ohai`
- `chef-sugar-rackspace`
- `sudo`
- `apt`

Usage
-----
#### rackspace_support::default
Just include `rackspace_support` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[rackspace_support]"
  ]
}
```

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
rackspace_support CHANGELOG
============================

1.1.0
-----
- [jarosser06] - Added firewall rules for new Rackspace bastions


1.0.0
-----
- [dude051] - Removed all firewall rules not directly related to Rackspace *Breaking change*

0.4.0
-----
- [jarosser06] - Added flag around firewall includes

0.3.0
-----
- [jarosser06] - Added tmux package
- [jarosser06] - Added systat default config for debian based distros

0.2.0
-----
- [jarosser06] - Added remote_ip tag

0.1.1
-----
- [jarosser06] - Allow Established connections

0.1.0
-----
- [jarosser06] - Initial release of rackspace_support


test/integration/default/serverspec/_firewall_spec.rb
# Encoding: utf-8

require 'spec_helper'

# Rackspace Bastion Access
%w(
  72.3.128.84/32
  69.20.0.1/32
  50.57.22.125/32
  120.136.34.22/32
  212.100.225.49/32
  212.100.225.42/32
  119.9.4.2/32
).each do |bast_range|
  describe iptables do
    it { should have_rule("-A INPUT -s #{bast_range} -p tcp -m tcp -m comment --comment \"Bastion Access\" -j ACCEPT") }
  end
end

# Rackspace Cloud Monitoring Access
%w(
  50.56.142.128/26
  50.57.61.0/26
  78.136.44.0/26
  180.150.149.64/26
  69.20.52.192/26
  92.52.126.0/24
).each do |cmon_range|
  describe iptables do
    it { should have_rule("-A INPUT -s #{cmon_range} -p tcp -m tcp -m comment --comment \"Cloud Monitoring Access\" -j ACCEPT") }
  end
end

# Rackspace Support Access
%w(
  50.56.230.0/24
  50.56.228.0/24
).each do |sup_range|
  describe iptables do
    it { should have_rule("-A INPUT -s #{sup_range} -p tcp -m tcp -m comment --comment \"Rackspace Support Access\" -j ACCEPT") }
  end
end


test/fixtures/cookbooks/rackspace_support_test/metadata.rb
name 'rackspace_support_test'
version '0.0.1'

depends 'rackspace_support'


test/fixtures/cookbooks/rackspace_support_test/recipes/_firewall.rb
#
# rackspace_support::_firewall fixture test
firewall 'iptables' do
  action :enable
  provider Chef::Provider::FirewallIptables
end

include_recipe 'rackspace_support::_firewall'


recipes/_ohai.rb
#
# Cookbook Name:: rackspace_support
# Recipe:: _ohai
#
# Copyright 2015, Rackspace
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Run all of this during compile time
include_recipe 'ohai'

directory node['ohai']['plugin_path'] do
  owner 'root'
  group 'root'
  mode 0755
  recursive true
end.run_action :create

remote_file ::File.join(node['ohai']['plugin_path'], 'rackconnect.rb') do
  source 'https://raw.githubusercontent.com/rackerlabs/ohai-plugins/master/plugins/rackconnect.rb'
  owner 'root'
  group 'root'
  mode 0755
  action :nothing
end.run_action :create

remote_file ::File.join(node['ohai']['plugin_path'], 'public_info.rb') do
  source 'https://raw.githubusercontent.com/rackspace-cookbooks/platformstack/master/files/default/public_info.rb'
  owner 'root'
  group 'root'
  mode 0755
  action :nothing
end.run_action :create

ohai('reload_ohai_for_rs').run_action :reload


recipes/_rack.rb
#
# Cookbook Name:: rackspace_support
# Recipe:: _rack
#
# Copyright 2015, Rackspace
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

group 'rack' do
  action :create
end

user 'rack' do
  gid 'rack'
  home '/home/rack'
  shell '/bin/bash'
  action :create
end

directory '/home/rack' do
  action :create
  mode 0755
  owner 'rack'
  group 'rack'
end

directory '/home/rack/.ssh' do
  action :create
  mode 0700
  owner 'rack'
  group 'rack'
end

remote_file '/home/rack/.ssh/authorized_keys' do
  source 'https://raw.github.com/rackops/authorized_keys/master/authorized_keys'
  owner 'rack'
  group 'rack'
  mode 0644
  use_conditional_get true
  use_etag true
  use_last_modified false
end

sudo 'rack' do
  user 'rack'
  nopasswd true
end


recipes/default.rb
#
# Cookbook Name:: rackspace_support
# Recipe:: default
#
# Copyright 2015, Rackspace
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

node.default['authorization']['sudo']['include_sudoers_d'] = true

%w(
  rackspace_support::_ohai
  rackspace_support::_common_pkgs
  sudo
  rackspace_support::_rack
  chef-sugar-rackspace
).each do |recipe|
  include_recipe recipe
end

if rackconnected?
  sudo 'rackconnect' do
    user 'rackconnect'
    nopasswd true
  end
end

remoteip = node.deep_fetch('public_info', 'remote_ip')
if remoteip =~ /[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+/
  # Assign the external_ip tag to the node if node['public_info']['remote_ip'] looks like an IP.
  tag("RemoteIP:#{remoteip}")
end


recipes/_common_pkgs.rb
#
# Cookbook Name:: rackspace_support
# Recipe:: _common_pkgs
#
# Copyright 2015, Rackspace
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

common_pkgs = %w(
  sysstat
  dstat
  screen
  curl
  telnet
  traceroute
  mtr
  zip
  lsof
  strace
  git
)

case node['platform_family']
when 'debian'
  common_pkgs.concat(%w(vim htop tmux))
  include_recipe 'apt'
when 'rhel'
  common_pkgs.push('vim-minimal')
end

common_pkgs.each do |pkg|
  package pkg do
    action :install
  end
end

# enable systat on Debian based platforms
file '/etc/default/sysstat' do
  only_if { node['platform_family'] == 'debian' }
  action :create
  content 'ENABLED="true"'
  mode 0644
  owner 'root'
  group 'root'
end


recipes/_firewall.rb
#
# Cookbook Name:: rackspace_support
# Recipe:: _firewall
#
# Copyright 2015, Rackspace
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Rackspace Bastion Access
%w(
  72.3.128.84/32
  69.20.0.1/32
  50.57.22.125/32
  120.136.34.22/32
  212.100.225.49/32
  212.100.225.42/32
  119.9.4.2/32
  172.99.99.10/32
  119.9.122.10/32
  146.20.2.10/32
  134.213.179.10/32
  134.213.178.10/32
  161.47.0.10/32
  119.9.148.10/32
).each do |bast_range|
  firewall_rule "bastion_#{bast_range}" do
    provider Chef::Provider::FirewallRuleIptables
    source bast_range
    action :allow
    description 'Bastion Access'
  end
end

# Rackspace Cloud Monitoring Access
%w(
  50.56.142.128/26
  50.57.61.0/26
  78.136.44.0/26
  180.150.149.64/26
  69.20.52.192/26
  92.52.126.0/24
).each do |cmon_range|
  firewall_rule "cloud_monitoring_#{cmon_range}" do
    provider Chef::Provider::FirewallRuleIptables
    source cmon_range
    action :allow
    description 'Cloud Monitoring Access'
  end
end

%w(
  50.56.230.0/24
  50.56.228.0/24
).each do |sup_range|
  firewall_rule "support_#{sup_range}" do
    provider Chef::Provider::FirewallRuleIptables
    source sup_range
    action :allow
    description 'Rackspace Support Access'
  end
end
